### DISCUSSION TITLE
Dict_Row for AsyncConnection.Execute / AsyncPool

### DISCUSSION THREAD

**Discussion by lacasaprivata2** [psycopg/psycopg#95] [General]
I was hoping there might be an example how to create an AsyncPool with Dict_Row as the row parser + calling fetchone() / fetchall() directly on the connection..... something like
```
with config.pool.connection() as conn:
  ostream = await conn.execute("QUERY").fetchall()
 ``` 
or
```
with config.pool.connection(row_parser=dict_row) as conn:
  ostream = await conn.execute("QUERY").fetchall()
```
To side track for a second, this request derives from the following problem:

When using CockroachDB, an increasingly popular DB provider grounded in serializable transactions, psycopg3 invokes transactions on the database on all cursors.....iIn async mode, for a machine w/ a small # of cores, any print() statements will cause many async functions to stop progress

Given the time it takes to print to the console and then resume coroutines, we end up extending beyond the database serializable transaction timeout limit causing an error

I'm hoping to overcome this problem by having Psycopg3 execute a SELECT / INSERT operation without a cursor akin to sqlx / asyncpg. Unfortunately, I don't think this is possible.

However, at the least, it would be great to have a shorter API where one can call execute() + fetchall() returning a list of dict_rows on an AsyncPool interface

**Comment by dvarrazzo** [comment]
Hi,

as per docs, the pool can take an `kwargs` parameter with the parameters to pass to the connection (https://www.psycopg.org/psycopg3/docs/api/pool.html#the-connectionpool-class). The connection takes a `row_factory` (https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.connect), so you can join the two with:

```python
from psycopg_pool import AsyncConnectionPool
from psycopg.row import dict_row
pool = AsyncConnectionPool(kwargs={"row_factory": dict_row})
```

The cursor is a thin wrapper around a result. It is a pure Python object and doesn't do any I/O, so I doubt it has any measurable impact on your performance. If it does please show a profile of that.

After you have a pool you can use it with (untested):

```python
async with pool.connection() as conn:
    cur = await conn.execute("select * from generate_series(1,3) n")
    records = await cur.fetchall()
```

you could use an `async with conn.cursor() as cur` but it's not necessary and adds some Python overhead for the context manager.

Everything else is something I don't understand: is it psycopg suspending the coroutine with some print or any other way, or it is your application?

**Comment by lacasaprivata2** [comment]
The print is within my application - a third party logger (stripe) prints to console synchronously w/out my control

It is important to note the database is CockroachDB - meaning it is not standard postgresql, albeit very close, and comes attached w/ a very picky transaction model for good reason given the strict serializability

The problem with not using the context manager is that, if i'm not mistaken, it won't rollback/auto-release the connection from the pool without it.

all said, it's still not clear to me why one can't have all this logic centralized w/ out a cursor & into the ```execute()``` command so that transactions are not needed at all and not started by default..... they are also much slower in my case. I understand that may or may not be compliant w/ the DB spec, the optionality on top of compliance makes a lot of sense

in summary, the problem here is only that all cursors start transactions. if async becomes stopped by blocking code, this poses a problem in a widely-deployed database CockroachDB

**Reply by dvarrazzo** [reply]
There is no relation between cursors and transactions. If you don't want transactions, use an autocommit connection. You can create a pool of autocommit connections passing `kwargs={"autocommit": True}` to the pool constructor.

I think you have the wrong mental model about how things work and you are doing pretty wild assumptions based on them. The way context managers work, who is in charge of transactions (hint: not the cursor) is explained in the documentation.

If you have a problem with a snippet of psycopg code please ask about it and try to solve one problem at time: the prints in your code or the use of CockroachDB are not something concerning this database adapter.
